assume-reflected definitions make some constraints fail

The following program fails verification.
```Haskell
{-@ LIQUID "--ple" @-}
{-@ LIQUID "--exactdc" @-}
module Test where

import Data.Set as Set

{-@ infixr ++ @-}

test :: Int -> ()
test x = lemma ([x] ++ [x]) (Set.singleton x)

{-@ lemma :: x:[Int] -> {y:_ | Set.isSubsetOf y (Set.listElts x) } -> { Set.isSubsetOf y (Set.listElts x) } @-}
lemma :: [Int] -> Set Int -> ()
lemma _ _ = ()

{-@ reflect append @-}
{-@ assume append :: xs:[a] -> ys:[a] -> {v:[a] | len v = len xs + len ys} @-}
append :: [a] -> [a] -> [a]
append [] ys = ys
append (x:xs) ys = x : append xs ys

{-@ assume reflect ++ as append @-}
```
Here's the error
```
Test.hs:10:29: error:
    Liquid Type Mismatch
    .
    The inferred type
      VV : {v : (Data.Set.Internal.Set GHC.Types.Int) | v == Set_sng x##aK5}
    .
    is not a subtype of the required type
      VV : {VV##1080 : (Data.Set.Internal.Set GHC.Types.Int) | Set_sub VV##1080 (Data.Set_LHAssumptions.listElts ?e)}
    .
    in the context
      ?d : {?d : [GHC.Types.Int] | ?d == GHC.Types.: x##aK5 ?c
                                   && Data.Set_LHAssumptions.listElts ?d == Set_cup (Set_sng x##aK5) (Data.Set_LHAssumptions.listElts ?c)
                                   && GHC.Types_LHAssumptions.len ?d == 1 + GHC.Types_LHAssumptions.len ?c
                                   && head ?d == x##aK5
                                   && lqdc##$select##GHC.Types.:##1 ?d == x##aK5
                                   && lqdc##$select##GHC.Types.:##2 ?d == ?c
                                   && tail ?d == ?c
                                   && GHC.Types_LHAssumptions.len ?d >= 0}
       
      ?c : {?c : [GHC.Types.Int] | Set_emp (Data.Set_LHAssumptions.listElts ?c)
                                   && ?c == GHC.Types.[]
                                   && GHC.Types_LHAssumptions.len ?c == 0
                                   && GHC.Types_LHAssumptions.len ?c >= 0}
       
      ?a : {?a : [GHC.Types.Int] | Set_emp (Data.Set_LHAssumptions.listElts ?a)
                                   && ?a == GHC.Types.[]
                                   && GHC.Types_LHAssumptions.len ?a == 0
                                   && GHC.Types_LHAssumptions.len ?a >= 0}
       
      ?b : {?b : [GHC.Types.Int] | ?b == GHC.Types.: x##aK5 ?a
                                   && Data.Set_LHAssumptions.listElts ?b == Set_cup (Set_sng x##aK5) (Data.Set_LHAssumptions.listElts ?a)
                                   && GHC.Types_LHAssumptions.len ?b == 1 + GHC.Types_LHAssumptions.len ?a
                                   && head ?b == x##aK5
                                   && lqdc##$select##GHC.Types.:##1 ?b == x##aK5
                                   && lqdc##$select##GHC.Types.:##2 ?b == ?a
                                   && tail ?b == ?a
                                   && GHC.Types_LHAssumptions.len ?b >= 0}
       
      ?e : {?e : [GHC.Types.Int] | GHC.Types_LHAssumptions.len ?e == GHC.Types_LHAssumptions.len ?b + GHC.Types_LHAssumptions.len ?d
                                   && GHC.Types_LHAssumptions.len ?e >= 0}
       
      x##aK5 : GHC.Types.Int
    Constraint id 8
   |
10 | test x = lemma ([x] ++ [x]) (Set.singleton x)
   |                             ^^^^^^^^^^^^^^^^^
```
The error message shows that `?e` isn't linked to `[x] ++ [x]` as it should. But verification passes if we replace the `++` call with `append`.
```diff
-test x = lemma ([x] ++ [x]) (Set.singleton x)
+test x = lemma ([x] `append` [x]) (Set.singleton x)
```

--- a/liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/Init.hs
+++ b/liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/Init.hs
@@ -82,7 +82,10 @@ initEnv info
        let tcb   = fmap (rTypeSort tce) <$> concat bs
        let cbs   = giCbs . giSrc $ info
        rTrue   <- mapM (traverse (true allowTC)) f6
-       let γ0    = measEnv sp (head bs) cbs tcb lt1s lt2s (f6 ++ bs!!3) (bs!!5) hs info
+       -- Issue #2537: reflected sigs (f6) must come AFTER assumed sigs (bs!!3) so that
+       -- when both exist for the same symbol (e.g., assume-reflect actual functions),
+       -- the strengthened reflected sig wins in M.fromList (which keeps the last entry).
+       let γ0    = measEnv sp (head bs) cbs tcb lt1s lt2s (bs!!3 ++ f6) (bs!!5) hs info
        γ  <- globalize <$> foldM (+=) γ0 ( [("initEnv", x, y) | (x, y) <- concat (rTrue:tail bs)])
        return γ {invs = is (invs1 ++ invs2)}
   where

--- a/tests/basic/pos/AssmRefl04.hs
+++ b/tests/basic/pos/AssmRefl04.hs
@@ -1,0 +1,26 @@
+{-@ LIQUID "--reflection" @-}
+{-@ LIQUID "--ple" @-}
+
+-- Test for issue #2537: assume-reflected definitions should not lose
+-- measure information when the actual function already has an assumed spec.
+
+module AssmRefl04 where
+
+import Data.Set as Set
+
+{-@ infixr ++ @-}
+
+test :: Int -> ()
+test x = lemma ([x] ++ [x]) (Set.singleton x)
+
+{-@ lemma :: x:[Int] -> {y:_ | Set.isSubsetOf y (Set.listElts x) } -> { Set.isSubsetOf y (Set.listElts x) } @-}
+lemma :: [Int] -> Set Int -> ()
+lemma _ _ = ()
+
+{-@ reflect append @-}
+{-@ assume append :: xs:[a] -> ys:[a] -> {v:[a] | len v = len xs + len ys} @-}
+append :: [a] -> [a] -> [a]
+append [] ys = ys
+append (x:xs) ys = x : append xs ys
+
+{-@ assume reflect ++ as append @-}

--- a/tests/tests.cabal
+++ b/tests/tests.cabal
@@ -2232,6 +2232,7 @@ executable basic-pos
                     , AssmRefl02C
                     , AssmRefl02D
                     , AssmRefl03
+                    , AssmRefl04
                     , AssmReflFilter
                     , DeadBind00
                     , Float
@@ -2284,6 +2285,7 @@ executable basic-pos
                     , TestAdmit
 
     build-depends:    base
+                    , containers
                     , liquid-prelude
                     , liquidhaskell
 

